File: scenarios/http/js/fetch/opentelemetry-fetch/test/client.test.js
// Copyright The OpenTelemetry Authors
// SPDX-License-Identifier: Apache-2.0

"use strict";

const assert = require("node:assert/strict");
const { spawn } = require("node:child_process");
const http = require("node:http");
const path = require("node:path");
const test = require("node:test");
const { URL } = require("node:url");

const grpc = require("@grpc/grpc-js");
const { respond } = require("@otel-conformance/http-test-client");

const TRACE_EXPORT_PATH =
  "/opentelemetry.proto.collector.trace.v1.TraceService/Export";

const service = {
  export: {
    path: TRACE_EXPORT_PATH,
    requestStream: false,
    responseStream: false,
    requestSerialize: (value) => value,
    requestDeserialize: (value) => value,
    responseSerialize: (value) => value,
    responseDeserialize: (value) => value,
  },
};

function listen(server) {
  return new Promise((resolve) => server.listen(0, "127.0.0.1", resolve));
}

function close(server) {
  server.closeAllConnections?.();
  return new Promise((resolve) => server.close(resolve));
}

function bind(server) {
  return new Promise((resolve, reject) => {
    server.bindAsync(
      "127.0.0.1:0",
      grpc.ServerCredentials.createInsecure(),
      (error, port) => (error ? reject(error) : resolve(port)),
    );
  });
}

function run(environment) {
  return new Promise((resolve, reject) => {
    const child = spawn(process.execPath, ["client.js"], {
      cwd: path.resolve(__dirname, ".."),
      env: { ...process.env, ...environment },
      stdio: "inherit",
    });
    child.once("error", reject);
    child.once("exit", (code) =>
      code === 0 ? resolve() : reject(new Error(`scenario exited ${code}`)),
    );
  });
}

function varint(buffer, offset) {
  let value = 0;
  let shift = 0;
  for (;;) {
    const byte = buffer[offset++];
    value += (byte & 0x7f) * 2 ** shift;
    if ((byte & 0x80) === 0) {
      return { offset, value };
    }
    shift += 7;
  }
}

function fields(buffer) {
  const result = [];
  let offset = 0;
  while (offset < buffer.length) {
    const key = varint(buffer, offset);
    offset = key.offset;
    const wireType = key.value & 7;
    const number = key.value >> 3;
    if (wireType === 0) {
      const value = varint(buffer, offset);
      result.push({ number, value: value.value, wireType });
      offset = value.offset;
      continue;
    }
    if (wireType === 2) {
      const size = varint(buffer, offset);
      offset = size.offset;
      result.push({
        number,
        value: buffer.subarray(offset, offset + size.value),
        wireType,
      });
      offset += size.value;
      continue;
    }
    if (wireType === 1 || wireType === 5) {
      const size = wireType === 1 ? 8 : 4;
      result.push({
        number,
        value: buffer.subarray(offset, offset + size),
        wireType,
      });
      offset += size;
      continue;
    }
    throw new Error(`unsupported protobuf wire type ${wireType}`);
  }
  return result;
}

function first(fieldsList, number, wireType) {
  return fieldsList.find(
    (field) => field.number === number && field.wireType === wireType,
  )?.value;
}

function all(fieldsList, number, wireType) {
  return fieldsList
    .filter((field) => field.number === number && field.wireType === wireType)
    .map((field) => field.value);
}

function text(value) {
  return value.toString("utf8");
}

function attribute(value) {
  const keyValue = fields(value);
  const key = text(first(keyValue, 1, 2));
  const anyValue = fields(first(keyValue, 2, 2));
  const raw = first(anyValue, 1, 2) ?? first(anyValue, 3, 0);
  return [key, Buffer.isBuffer(raw) ? text(raw) : raw];
}

function decodeSpans(payload) {
  return all(fields(payload), 1, 2).flatMap((resourceSpans) =>
    all(fields(resourceSpans), 2, 2).flatMap((scopeSpans) =>
      all(fields(scopeSpans), 2, 2).map((encoded) => {
        const span = fields(encoded);
        return {
          attributes: new Map(all(span, 9, 2).map(attribute)),
          kind: first(span, 6, 0),
          name: text(first(span, 5, 2)),
        };
      }),
    ),
  );
}

test("measures only the selected browser Fetch request", async () => {
  const mock = http.createServer(async (request, response) => {
    const chunks = [];
    for await (const chunk of request) {
      chunks.push(chunk);
    }
    const body = chunks.length ? Buffer.concat(chunks).toString("utf8") : null;
    const answer = respond(
      request.method,
      new URL(request.url, "http://unused").pathname,
      body,
    );
    response.writeHead(answer.status, {
      "content-type": "application/json",
    });
    response.end(answer.body);
  });
  await listen(mock);

  const exports = [];
  const collector = new grpc.Server();
  collector.addService(service, {
    export(call, callback) {
      exports.push(call.request);
      callback(null, Buffer.alloc(0));
    },
  });
  const collectorPort = await bind(collector);
  collector.start();

  try {
    await run({
      MOCK_SERVER_URL: `http://127.0.0.1:${mock.address().port}`,
      OTEL_CONFORMANCE_SCENARIO_INDEX: "2",
      OTEL_EXPORTER_OTLP_ENDPOINT: `http://127.0.0.1:${collectorPort}`,
    });
  } finally {
    await close(mock);
    collector.forceShutdown();
  }
  const spans = exports.flatMap(decodeSpans);
  assert.equal(spans.length, 1, "the exporter request must not create a span");
  assert.deepEqual(
    spans.map(({ attributes, kind, name }) => {
      const url = new URL(attributes.get("url.full"));
      return {
        kind,
        method: attributes.get("http.request.method"),
        name,
        path: `${url.pathname}${url.search}`,
        status: attributes.get("http.response.status_code"),
      };
    }),
    [
      {
        kind: 3,
        method: "POST",
        name: "POST",
        path: "/items",
        status: 201,
      },
    ],
  );
});

File: scenarios/http/ruby/net_http/opentelemetry-instrumentation-net_http/client/test/client_test.rb
# frozen_string_literal: true

# Copyright The OpenTelemetry Authors
# SPDX-License-Identifier: Apache-2.0

require "minitest/autorun"
require_relative "../../../scenarios/client"

class NetHttpScenarioClientTest < Minitest::Test
  HTTPContract = OpenTelemetry::Conformance::HTTP
  ACTIONS = '[{"request":{"method":"GET","path":"/health"},"response":{"status":200,"body":"{\"ok\": true}"}},{"request":{"method":"POST","path":"/items","body":"{\"name\": \"widget\"}"},"response":{"status":201,"body":"{\"created\": true, \"payload\": ${requestBody}}"}},{"request":{"method":"GET","path":"/status/500"},"response":{"status":500,"body":"{\"message\": \"status 500\"}"}}]'.freeze
  ACTION = '{"request":{"method":"POST","path":"/items","body":"{\"name\": \"widget\"}"},"response":{"status":201,"body":"{\"created\": true, \"payload\": ${requestBody}}"}}'.freeze

  Response = Struct.new(:code, :body)

  class FakeNetHTTP
    class << self
      attr_reader :instances

      def reset
        @instances = []
      end
    end

    attr_accessor :use_ssl
    attr_reader :host, :port, :requests, :starts

    def initialize(host, port)
      @host = host
      @port = port
      @requests = []
      @starts = 0
      self.class.instances << self
    end

    def start
      @starts += 1
      yield self
    end

    def request(request)
      @requests << request
      answer = HTTPContract.respond(request.method, request.path, request.body)
      Response.new(answer.status.to_s, answer.body)
    end
  end

  def setup
    FakeNetHTTP.reset
    ENV[HTTPContract::ACTIONS_VARIABLE] = ACTIONS
    ENV[HTTPContract::ACTION_VARIABLE] = ACTION
  end

  def test_runs_the_contract_over_one_connection
    capture_io do
      NetHttpScenario.run(
        "https://example.test:8443",
        http_class: FakeNetHTTP
      )
    end

    assert_equal 1, FakeNetHTTP.instances.length
    connection = FakeNetHTTP.instances.fetch(0)
    assert_equal "example.test", connection.host
    assert_equal 8443, connection.port
    assert_equal true, connection.use_ssl
    assert_equal 1, connection.starts
    assert_equal 1, connection.requests.length
    assert_equal ["/items"], connection.requests.map(&:path)
  end

  def test_sends_the_shared_headers_and_body
    capture_io do
      NetHttpScenario.run("http://example.test", http_class: FakeNetHTTP)
    end

    requests = FakeNetHTTP.instances.fetch(0).requests
    requests.each do |request|
      assert_equal HTTPContract::USER_AGENT, request["User-Agent"]
    end

    post = requests.find { |request| request.method == "POST" }
    assert_equal HTTPContract::CONTENT_TYPE, post["Content-Type"]
    assert_equal '{"name": "widget"}', post.body

    requests.reject { |request| request.equal?(post) }.each do |request|
      assert_nil request["Content-Type"]
      assert_nil request.body
    end
  end

  def test_returns_helper_responses_for_error_statuses
    client = NetHttpScenario::Client.new(
      "http://example.test",
      http_class: FakeNetHTTP
    )

    response = client.request(
      "GET",
      "http://example.test/status/500",
      nil
    )

    assert_instance_of HTTPContract::Response, response
    assert_equal 500, response.status
    assert_equal '{"message": "status 500"}', response.body
  end
end
